### PR TITLE
Adding IntelliSense support using JSDoc

### PR DESCRIPTION
**Pull request by rahul799**
This PR will leverage the IntelliSense support. There are still some Glitches in the documentation, which will be improved in the subsequent PR's.

### CODE REVIEW

**Comment by maxkfranz** [comment]
@rahul799, could you modify this PR so that it would merge to the [typescript](https://github.com/cytoscape/cytoscape.js/tree/typescript) branch instead of the unstable branch?

**Comment by maxkfranz** [comment]
It also looks like the Travis build failed due to some missing dependencies.

**Comment by maxkfranz** [comment]
> @rahul799, could you modify this PR so that it would merge to the [typescript](https://github.com/cytoscape/cytoscape.js/tree/typescript) branch instead of the unstable branch?

It looks like I misread the branches line.  Disregard the above comment.

**Comment by rahul799** [comment]
Hope it fixes Travis :crossed_fingers:

File: documentation/generateDocs.js [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+// save generated file
+fs.writeFile ( path.join(__dirname, "./docmaker.json"), JSON.stringify(dockmaker_elements, null,4), function(err) {
```
Comment (maxkfranz) [inline documentation/generateDocs.js:271]: If you use `fs.writeFileSync()` everywhere you use `fs.writeFile()`, you won't need the sleep delay.

File: documentation/generateDocs.js [hunk 2]
```diff
@@ -0,0 +1,275 @@
+var fs= require('fs');
+var path = require('path');
+var data=fs.readFileSync( path.join(__dirname, './docmaker.json'), 'utf8');
+var dockmaker_elements=JSON.parse(data);
+
+const { exec } = require('child_process');
+exec('jsdoc -X ./src/[c,e,u]*/*.js > ./documentation/AST/core_AST.json && jsdoc -X ./src/animation.js > ./documentation/AST/animation_AST.json && jsdoc -X ./src/extensions/layout/index.js > ./documentation/AST/layout_AST.json', { "shell": "/bin/bash" },(err, stdout, stderr) => {
```
Comment (maxkfranz) [inline documentation/generateDocs.js:7]: This exec should be a separate npm script (or several scripts) in package.json.  The npm scripts should run before `generateDocs.js`.

That way, the docs can be built on Windows as well as Linux.

File: documentation/generateDocs.js (hunk 1 above)
Comment (rahul799) [reply documentation/generateDocs.js]: Fixed.

File: documentation/generateDocs.js (hunk 2 above)
Comment (rahul799) [reply documentation/generateDocs.js]: Thank you for pointing it out. I have fixed it.

**Comment by rahul799** [comment]
Travis is falling with the error log `linkify-it not accessible from markdown-it`. @maxkfranz any Idea what is causing the issue?

**Comment by maxkfranz** [comment]
It's probably another missing dependency.  Remove your local `node_modules` directory, run `npm install` again, and you should be able to reproduce the error locally.

**Comment by maxkfranz** [comment]
OK, it looks like Travis is passing the tests now.  I took a bit of a deeper look into the PR, and I have some feedback.

In `package.json`:

(1) `tsd-jsdoc` should be a dev dependency, not a regular dependency.

(2) Docs-related scripts should all use the `docs:` prefix.  How about `docs:ast` (make the AS) and `docs:json` (generate docmaker.json)?  The `generateDocs.js` file might be better named `ast-to-docmaker.js`.

(3) Generally, `package.json` scripts and JS file names should be in lowercase.

Running:

(4) I tried clearing out the `fns` to `fns: []` under `Graph manipulation` to test that the `docmaker.json` file would be rebuilt by running `npm run docs`.  The file was not updated, and `fns` remained empty.

(5) Where is the npm script for generating the typescript defintions?

Cleanup:

(6) The `documentation/AST` directory and its containing files should be in lowercase.  The files in the `documentation/ast` directory should be listed as entries in `.gitignore`.

(7) `generateAST.js` should be one or more npm scripts in `package.json`.  Each one of these is a separate script:

- docs:ast:core: `jsdoc -X ./src/[c,e,u]*/*.js > ./documentation/AST/core_AST.json`
- docs:ast:animation: `jsdoc -X ./src/animation.js > ./documentation/AST/animation_AST.json`
- docs:ast:layout: `jsdoc -X ./src/extensions/layout/index.js > ./documentation/AST/layout_AST.json`
- Is collection missing?
- You can use `run-s docs:ast:*` for `docs:ast`.  See the package.json for examples, and refer to the docs for [npm-run-all](https://www.npmjs.com/package/npm-run-all).

(8) Should `jsdocAST.json` be under `documentation/ast`?

**Comment by rahul799** [comment]
Not stale

**Pull request merged**